src/Features/Core/Portable/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+                        cancellationToken, out var targetArgument, out var targetArgumentConversionType)
+                    && GetArgumentExpression(targetArgument) is TExpressionSyntax argumentExpression)
+                {
+                    mutablePotentialConversionTypes.Add(Tuple.Create(argumentExpression, targetArgumentConversionType));
+                }
+            }
+
+            // Sort the potential conversion types by inheritance distance, so that
+            // operations are in order and user can choose least specific types(more accurate)
+            mutablePotentialConversionTypes.Sort(new InheritanceDistanceComparer<TExpressionSyntax>(semanticModel));
+
+            return mutablePotentialConversionTypes.ToImmutable();
+        }
+
+        protected bool FindCorrespondingParameterByName(
+            string argumentName, ImmutableArray<IParameterSymbol> parameters, ref int parameterIndex)
+        {
+            for (var j = 0; j < parameters.Length; j++)
+            {
+                if (argumentName.Equals(parameters[j].Name))
+                {
+                    parameterIndex = j;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var spanNodes = diagnostics.SelectAsArray(
+                d => root.FindNode(d.Location.SourceSpan, getInnermostNodeForTie: true)
+                         .GetAncestorsOrThis<TExpressionSyntax>().FirstOrDefault());
+
+            await editor.ApplyExpressionLevelSemanticEditsAsync(
+                document, spanNodes,
+                (semanticModel, spanNode) => true,
+                (semanticModel, currentRoot, spanNode) =>
+                {
+                    // All diagnostics have the same error code
+                    if (TryGetTargetTypeInfo(document, semanticModel, currentRoot, diagnostics[0].Id, spanNode,
+                        cancellationToken, out var potentialConversionTypes)
+                        && potentialConversionTypes.Length == 1)
+                    {
+                        return ApplyFix(currentRoot, potentialConversionTypes[0].Item1, potentialConversionTypes[0].Item2);
+                    }
+
+                    return currentRoot;
+                },
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument, equivalenceKey: title)
+            {
+            }
+        }
+    }
```

Comment:
can you let me know if any of these functions changed in interesting ways?  or if they just moved up to the new abstract base class?

Reply:
I feel like the majority changes are just to move up the functions to abstract base class. The main change beside this is that `TryGetTargetTypeInfo` outputs a list of `(node, conversionType)` instead of a list of `conversionType`, and sub functions changed correspondingly